### PR TITLE
Create e-puck2 driver

### PR DESCRIPTION
**Pull request by lukicdarkoo**
**Description**
ROS2 e-puck2 driver for Webots should be created. The API should match to one implemented for `epuck_ros2` (https://github.com/cyberbotics/epuck_ros2).

**Tasks**
  - [x] Support for camera (Image and CameraInfo).
  - [x] Add Odometry topic
  - [x] Support for LEDs.
  - [x] Add Twist topic.
  - [x] Add LaserScan topic.
  - [x] Add Imu topic.
  - [x] Publish corresponding transformations.
  - [x] Match LED API with `epuck_ros2` 
  - [x] Add support for light sensors
  - [x] Add support for ground sensors if present
  - [x] Publish static transforms
    - [x] ground sensors,
    - [x] distance sensors,
    - [x] light sensors and
    - [x] laser scanner.
  - [x] Verify compatibility with `epuck_ros2` API
  - [x] Verify everything in simulation
    - [x] Odometry + Twist
    - [x] Camera + CameraInfo
    - [x] LaserScan + Range
    - [x] Ground sensors
    - [x] Light sensors
  - [x] Make gyroscope, ground sensors and ToF optional to keep backwards compatibility

### CODE REVIEW

**Comment by lukicdarkoo** [comment]
image
Output from `rqt_graph`

**Comment by DavidMansolino** [comment]
👀

File: webots_ros2_epuck2/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/cmd_vel [geometry_msgs/msg/Twist]
+/distance/ps0 [sensor_msgs/msg/Range]
+/distance/ps1 [sensor_msgs/msg/Range]
+/distance/ps2 [sensor_msgs/msg/Range]
+/distance/ps3 [sensor_msgs/msg/Range]
+/distance/ps4 [sensor_msgs/msg/Range]
+/distance/ps5 [sensor_msgs/msg/Range]
+/distance/ps6 [sensor_msgs/msg/Range]
+/distance/ps7 [sensor_msgs/msg/Range]
+/distance/tof [sensor_msgs/msg/Range]
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/README.md]: It seems we are missing the ground sensors, if they are present (``E-puckGroundSensors`` node in the ``groundSensorsSlot`` field). But this can go in another PR of course.

File: webots_ros2_epuck2/launch/example_launch.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+
+# Copyright 1996-2019 Cyberbotics Ltd.
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/launch/example_launch.py]:
```suggestion
# Copyright 1996-2020 Cyberbotics Ltd.
```

File: webots_ros2_epuck2/launch/example_launch.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# ros2 launch webots_ros2_epuck2 example_launch.py
+
+"""Launch Webots and the controller."""
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/launch/example_launch.py]:
```suggestion
"""Launch Webots, the controller and Rviz."""
```

File: webots_ros2_epuck2/package.xml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  <exec_depend>rclpy</exec_depend>
+  <exec_depend>std_msgs</exec_depend>
+  <exec_depend>nav_msgs</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
+
+  <exec_depend>geometry_msgs</exec_depend>
+  <exec_depend>webots_ros2_msgs</exec_depend>
+  <exec_depend>builtin_interfaces</exec_depend>
+  <exec_depend>tf2_ros</exec_depend>
+  <exec_depend>webots_ros2_core</exec_depend>
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/package.xml]: We have a dependency on Rviz too due to the launch file.

File: webots_ros2_epuck2/resource/epuck2_world.wbt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    Pi-puck {
+    }
+  ]
+}
+DirectionalLight {
+  direction 0.5 -1 0.5
+  intensity 1.8
+  castShadows TRUE
+  castLensFlares TRUE
+}
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/resource/epuck2_world.wbt]: It is better to use instead the ``TexturedBackgroundLight``

File: webots_ros2_epuck2/setup.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    maintainer='Cyberbotics',
+    maintainer_email='support@cyberbotics.com',
+    keywords=['ROS', 'Webots', 'Robot', 'Simulation', 'Examples'],
+    classifiers=[
+        'Intended Audience :: Developers',
+        'License :: OSI Approved :: Apache Software License',
+        'Programming Language :: Python',
+        'Topic :: Software Development',
+    ],
+    description='Minimal example showing how to control a robot with ROS2 in Webots.',
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/setup.py]: This should be updated.

File: webots_ros2_epuck2/test/wb_odom_test.py.txt [hunk 7]
```diff
@@ -0,0 +1,99 @@
+"""odom_test controller."""
+
+# You may need to import some classes of the controller module. Ex:
+#  from controller import Robot, Motor, DistanceSensor
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/test/wb_odom_test.py.txt]:
```suggestion
```

File: webots_ros2_epuck2/webots_ros2_epuck2/__init__.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Copyright 1996-2019 Cyberbotics Ltd.
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/__init__.py]:
```suggestion
# Copyright 1996-2020 Cyberbotics Ltd.
```

File: webots_ros2_epuck2/worlds/epuck2_world.wbt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+}
+E-puck {
+  version "2"
+  controller "<extern>"
+  controllerArgs ""
+  turretSlot [
+    Pi-puck {
+    }
+  ]
+}
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/worlds/epuck2_world.wbt]: The world file is duplicated, once here and once in the resources.

File: webots_ros2_epuck2/test/wb_odom_test.py.txt [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+right_motor.setVelocity(-0.9)
+
+steps = 700
+while robot.step(timestep) != -1:
+    test.odometry_callback()
+    if steps < 0:
+        left_motor.setVelocity(0)
+        right_motor.setVelocity(0)
+    else:
+        steps -= 1
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/test/wb_odom_test.py.txt]: I am not sure to understand the purpose of this test.
It is not using the ROS nodes (i.e. driver), and is it called from the CI?

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    t1 = +1.0 - 2.0 * (q.x * q.x + q.y * q.y)
+    roll = math.atan2(t0, t1)
+    t2 = +2.0 * (q.w * q.y - q.z * q.x)
+    t2 = +1.0 if t2 > +1.0 else t2
+    t2 = -1.0 if t2 < -1.0 else t2
+    pitch = math.asin(t2)
+    t3 = +2.0 * (q.w * q.z + q.x * q.y)
+    t4 = +1.0 - 2.0 * (q.y * q.y + q.z * q.z)
+    yaw = math.atan2(t3, t4)
+    return [yaw, pitch, roll]
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: We might move this generic function in a ``math_utils``file in this package to make it available to all the packages: https://github.com/cyberbotics/webots_ros2/tree/master/webots_ros2_core/webots_ros2_core

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import rclpy
+import math
+from nav_msgs.msg import Odometry
+from rclpy.node import Node
+from rcl_interfaces.srv import SetParameters
+from rcl_interfaces.msg._parameter import Parameter
+from rclpy.parameter import ParameterType, ParameterValue
+from geometry_msgs.msg import Twist
+
+
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: A small introcution comment explaining what does this script would be welcome.

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        self.create_subscription(Odometry, '/odom', self.odometry_callback, 1)
+        self.pub = self.create_publisher(Twist, '/cmd_vel', 10)
+
+        # Parameter service
+        self.cli = self.create_client(SetParameters, 'epuck2/set_parameters')
+        self.cli.wait_for_service(timeout_sec=1.0)
+        self.set_param('wheel_distance', self.wheel_distance.value)
+        self.set_param('wheel_radius', self.wheel_radius.value)
+        print('Setting wheel distance to: {}m'.format(self.wheel_distance.value))
+        print('Setting wheel radius to: {}m'.format(self.wheel_radius.value))
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: This should probably use the ROS log system.

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+            # Receive data
+            yaw, _, _ = quaternion_to_euler(msg.pose.pose.orientation)
+            if yaw > 0 and self.last_yaw < 0:
+                self.rotation_count += 1
+            if self.rotation_count == 4:
+                self.send_stop()
+            self.last_yaw = yaw
+
+            print('Circle: {}; Current angle: {}'.format(
+                self.rotation_count, yaw))
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: This should probably use the ROS log system.

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+        if not self.test_done and self.type.value == 'linear':
+            # Send velocity
+            print('Rotation calibration in progress...')
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: This should probably use the ROS log system.

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            msg_twist.angular.x = 0.0
+            msg_twist.angular.y = 0.0
+            msg_twist.angular.z = 0.0
+            msg_twist.linear.x = 0.02
+            msg_twist.linear.y = 0.0
+            msg_twist.linear.z = 0.0
+            self.pub.publish(msg_twist)
+
+            # Receive data
+            print('X', msg.pose.pose.position.x)
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: This should probably use the ROS log system.

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(self, name, args=None):
+        super().__init__(name)
+
+        self.test_done = False
+        self.rotation_count = 0
+        self.last_yaw = 0
+
+        # Parameters
+        self.type = self.declare_parameter('type', 'rotation')
+        self.distance = self.declare_parameter('distance', 133.5 / 1000)
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: Where does this 233.5 value comes from?

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        msg.angular.z = 0.0
+        msg.linear.x = 0.0
+        msg.linear.y = 0.0
+        msg.linear.z = 0.0
+        self.pub.publish(msg)
+
+    def odometry_callback(self, msg: Odometry):
+        if not self.test_done and self.type.value == 'rotation':
+            # Send velocity
+            print('Rotation calibration in progress...')
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: This should probably use the ROS log system.

File: webots_ros2_epuck2/README.md [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+/imu [sensor_msgs/msg/Imu]
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/README.md]: ``ls0-7``are missing.

File: webots_ros2_epuck2/package.xml [hunk 20]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
+  <name>webots_ros2_epuck2</name>
+  <version>0.0.2</version>
+  <description>Minimal example showing how to control a robot with ROS2 in Webots.</description>
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/package.xml]: This should be updated.

File: webots_ros2_epuck2/README.md (hunk 19 above)
Comment (DavidMansolino) [reply webots_ros2_epuck2/README.md]: The ``scan``too.

File: webots_ros2_epuck2/webots_ros2_epuck2/driver.py [hunk 21]
```diff
@@ -0,0 +1,617 @@
...
+        )
+    else:
+        # Interpolate as last
+        return interpolate_function(
+            value,
+            table[len(table) - 2][1],
+            table[len(table) - 2][0],
+            table[len(table) - 1][1],
+            table[len(table) - 1][0]
+        )
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/webots_ros2_epuck2/driver.py]: This function too might be moved in a ``math_utils`` file in this package to make it available to all the packages: https://github.com/cyberbotics/webots_ros2/tree/master/webots_ros2_core/webots_ros2_core

File: webots_ros2_epuck2/worlds/.epuck2_world.wbproj [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+perspectives: 000000ff00000000fd00000003000000000000000000000000fc0100000001fb0000001a0044006f00630075006d0065006e0074006100740069006f006e0000000000ffffffff0000006900ffffff00000001000001780000031cfc0200000001fb0000001400540065007800740045006400690074006f007201000000160000031c0000008900ffffff0000000300000453000000b7fc0100000001fb0000000e0043006f006e0073006f006c00650100000000000004530000006900ffffff000002d90000031c00000001000000020000000100000008fc00000000
+simulationViewPerspectives: 000000ff000000010000000200000137000001a00100000002010000000100
+sceneTreePerspectives: 000000ff0000000100000002000000c0000000fa0100000002010000000200
+maximizedDockId: -1
+centralWidgetVisible: 1
+selectionDisabled: 0
+viewpointLocked: 0
+orthographicViewHeight: 1
+textFiles: 0 "controllers/4_wheels_collision_avoidance/4_wheels_collision_avoidance.py"
+globalOptionalRendering: ContactPoints::CoordinateSystem::DistanceSensorRays
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/worlds/.epuck2_world.wbproj]:
```suggestion
globalOptionalRendering: DistanceSensorRays
```

File: webots_ros2_epuck2/worlds/.epuck2_world.wbproj [hunk 23]
```diff
@@ -0,0 +1,12 @@
+Webots Project File version R2020a
+perspectives: 000000ff00000000fd00000003000000000000000000000000fc0100000001fb0000001a0044006f00630075006d0065006e0074006100740069006f006e0000000000ffffffff0000006900ffffff00000001000001780000031cfc0200000001fb0000001400540065007800740045006400690074006f007201000000160000031c0000008900ffffff0000000300000453000000b7fc0100000001fb0000000e0043006f006e0073006f006c00650100000000000004530000006900ffffff000002d90000031c00000001000000020000000100000008fc00000000
+simulationViewPerspectives: 000000ff000000010000000200000137000001a00100000002010000000100
+sceneTreePerspectives: 000000ff0000000100000002000000c0000000fa0100000002010000000200
+maximizedDockId: -1
+centralWidgetVisible: 1
+selectionDisabled: 0
+viewpointLocked: 0
+orthographicViewHeight: 1
+textFiles: 0 "controllers/4_wheels_collision_avoidance/4_wheels_collision_avoidance.py"
```
Comment (DavidMansolino) [inline webots_ros2_epuck2/worlds/.epuck2_world.wbproj]:
```suggestion
textFiles: 0
```

File: webots_ros2_epuck2/resource/epuck2_world.wbt (hunk 5 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/resource/epuck2_world.wbt]: I started with `TexturedBackgroundLight` but we cannot present functionality of light sensors with that kind of lighting - the illumination detected by the sensors is at maximum even in a shadow

File: webots_ros2_epuck2/README.md (hunk 1 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/README.md]: Added more detailed introduction

File: webots_ros2_epuck2/README.md (hunk 19 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/README.md]: Detailed introduction added

File: webots_ros2_epuck2/resource/epuck2_world.wbt (hunk 5 above)
Comment (DavidMansolino) [reply webots_ros2_epuck2/resource/epuck2_world.wbt]: Ok, makes sense, in that case you can indeed keep this one.

File: webots_ros2_epuck2/webots_ros2_epuck2/driver.py (hunk 21 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/webots_ros2_epuck2/driver.py]: I also moved `get_ros_stamp()` to `utils.py`, please check if it is a proper place for it

Comment (DavidMansolino) [reply webots_ros2_epuck2/webots_ros2_epuck2/driver.py]: Yes, that's indeed good to move it there.

File: webots_ros2_epuck2/test/wb_odom_test.py.txt (hunk 10 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/test/wb_odom_test.py.txt]: I used that controller directly as Webots controller to test odometry. But it should be deleted as it will not be of help to anybody else in the future

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py (hunk 13 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: It is indeed better, replaced

File: webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py (hunk 17 above)
Comment (lukicdarkoo) [reply webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: It was integer number of like 10 rectangles in my notebook :) I will define it properly

Comment (lukicdarkoo) [reply webots_ros2_epuck2/webots_ros2_epuck2/drive_calibrator.py]: Resolved

**Comment by lukicdarkoo** [comment]
@DavidMansolino Should we change a name from `webots_ros2_epuck2` to `webots_ros2_epuck`?

**Comment by DavidMansolino** [comment]
> @DavidMansolino Should we change a name from `webots_ros2_epuck2` to `webots_ros2_epuck`?

That's a good idea indeed, but in that case, please make sure that it is working with the e-puck 1 too (the controller should handle the missing devices such as ToF and Gyro).

**Comment by lukicdarkoo** [comment]
> > @DavidMansolino Should we change a name from `webots_ros2_epuck2` to `webots_ros2_epuck`?
> 
> That's a good idea indeed, but in that case, please make sure that it is working with the simulated e-puck 1 too (the controller should handle the missing devices such as ToF and Gyro).

Well, gyro supposed to be available in e-puck1 as well according to documentation. It seems it is misses in the model. Here is a comparison between e-puck1 and e-puck2:
https://www.gctronic.com/doc/index.php/e-puck2
in e-puck2 3D magnetometer is added. Btw, the magnetometer also seems to be missing in e-puck2 proto, maybe I should add. As well as the support for the sensor in the physical robot.

Anyway, I will make everything optional to preserve backwards compatibility.

**Comment by lukicdarkoo** [comment]
> Note also that to make this code easily re-usable and make it more generic I would like that you create a `WebotsDifferentialWheelNode` (or any better name) that inherits from `WebotsNode` and is in the `webots_ros2_core` package and provides all the functionalities related to the odom publisher and cmd_vel subscriber (the `EPuckDriver` class will then inherit from this class).
> 
> This way it will be very simple to create new differential wheels robot nodes based on this excellent code you did. For example the `webots_ros2_tiago`could use it.
> But please don't do this in this PR as it is already very good and functional like this (and start to be quite big ).

I love the idea! I will do it! Gazebo is doing something similar (http://gazebosim.org/tutorials?tut=ros_gzplugins) but this could be a cleaner and more configurable approach.

File: webots_ros2_epuck/webots_ros2_epuck/driver.py [hunk 24]
```diff
@@ -0,0 +1,569 @@
...
+            ground_sensor = self.robot.getDistanceSensor(idx)
+            if ground_sensor:
+                ground_sensor.enable(self.period.value)
+                self.ground_sensors[idx] = ground_sensor
+                self.ground_sensor_publishers[idx] = self.create_publisher(
+                    Range, '/' + idx, 1)
+
+                ground_sensor_broadcaster = StaticTransformBroadcaster(self)
+                ground_sensor_transform = TransformStamped()
+                ground_sensor_transform.header.stamp = get_ros_stamp()
```
[the remark this answers is not in the document]
Comment (lukicdarkoo) [reply webots_ros2_epuck/webots_ros2_epuck/driver.py]: Yes! Added to `webots_ros2_core`
